contracts/token/Basic223Token.sol

```diff
@@ -0,0 +1,66 @@
+pragma solidity ^0.4.18;
+
+import './ERC223Basic.sol';
+import './ERC223Receiver.sol';
+import './BasicToken.sol';
+
+ /**
+  *
+  * @title Basic token ERC223 
+  *        derived from Basic token ERC20
+  * @dev see also: https://github.com/ethereum/EIPs/issues/223  
+  *
+  * created by IAM <DEV> (Elky Bachtiar) 
+  * https://www.iamdeveloper.io
+  *
+  *
+  * file: Basic223Token.sol
+  * location: contracts/token/
+  *
+ */
+contract Basic223Token is ERC223Basic, BasicToken {
+  
+    /**
+    * @dev transfer token for a specified address
+    * @param _to The address to transfer to.
+    * @param _value The amount to be transferred
+    * @param _data is arbitrary data sent with the token transferFrom. Simulates ether tx.data
+    * @return bool successful or not
+    */
+    function transfer(address _to, uint _value, bytes _data) public returns (bool success) {
+        require(_to != address(0));
+        require(_value <= balances[msg.sender]);
+        require(balances[_to].add(_value) > balances[_to]);  // Detect balance overflow
+    
+        assert(super.transfer(_to, _value));               //@dev Save transfer
+
+        if (isContract(_to)){
+          return contractFallback(msg.sender, _to, _value, _data);
+        }
+        return true;
+    }
+
+    /**
+    * @dev transfer token for a specified address
+    * @param _to The address to transfer to.
+    * @param _value The amount to be transferred.
+    */
+    function transfer(address _to, uint256 _value) public returns (bool success) {        
+        return transfer(_to, _value, new bytes(0));
+    }
+
+
+    //function that is called when transaction target is a contract
+    function contractFallback(address _origin, address _to, uint _value, bytes _data) internal returns (bool success) {
+        ERC223Receiver reciever = ERC223Receiver(_to);
+        return reciever.tokenFallback(msg.sender, _origin, _value, _data);
```

Comment:
I think the implementation here is wrong since a contract receiver has no chance to revert when someone is sending tokens to it. delegatecall(new bytes(0)) just return in tokenFallback(msg.sender, _origin, _value, new bytes(0)) and the transaction will success, and then these tokens will loss. An implementation by @Dexaran [ERC233](https://github.com/Dexaran/ERC223-token-standard/blob/master/token/ERC223/ERC223_token.sol) is more reasonable since token.transfer() call receiver.tokenFailback() directly and the receiver could have a chance to revert if sending tokens to the receiver's address is not what it expected.

Reply:
a workaround is to check _data not bytes(0) in tokenFallback.
if (_data == bytes(0)) revert;

Reply:
or maybe:
if (_data == bytes(0)) defaultTokenPayableAction();